### PR TITLE
[Improve][Connector-V2][JDBC] Unified exception for JDBC source & sink

### LINKED ISSUE #3556: [Improve][Connector-V2][JDBC] Unified exception for JDBC source & sink connector
sub-task of #2930

### PR DESCRIPTION
**Pull request by Carl-Zhou-CN**
…k connector

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

close #3556

### CODE REVIEW

**Comment by Carl-Zhou-CN** [comment]
@TyrantLucifer @EricJoy2048 @Hisoka-X  PTAL.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+package org.apache.seatunnel.connectors.seatunnel.jdbc.exception;
+
+import org.apache.seatunnel.common.exception.SeaTunnelErrorCode;
+
+public enum JdbcConnectorErrorCode implements SeaTunnelErrorCode {
+
+    CREATE_DRIVER_FAILED("JDBC-01", "Fail to create driver of class"),
+    NO_SUITABLE_DRIVER("JDBC-02", "No suitable driver found"),
+    SQL_OPERATION_FAILED("JDBC-03", "Sql operation failed, such as (execute,addBatch,close) etc..."),
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java]: Remove useless error code.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    XA_EXCEPTION("JDBC-04", "Xa operation failed, such as (commit, rollback) etc.."),
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java]:
```suggestion
    XA_OPERATION_FAILED("JDBC-04", "Xa operation failed, such as (commit, rollback) etc.."),
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    CONNECT_DATABASE_FAILED("JDBC-05", "Connector database failed"),
+    NULL_POINTER("JDBC-06", "Null pointer failed"),
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java]: This error code is not clearly for users. Please change it.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    EXCEPTION("JDBC-07", "transaction operation failed, such as (commit, rollback) etc.."),
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java]:
```suggestion
    TRANSACTION_OPERATION_FAILED("JDBC-07", "transaction operation failed, such as (commit, rollback) etc.."),
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    private final String description;
+
+    JdbcConnectorErrorCode(String code, String description) {
+        this.code = code;
+        this.description = description;
+    }
+
+    @Override
+    public String getCode() {
+        return null;
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java]: return code

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    @Override
+    public String getDescription() {
+        return null;
```
Comment (TyrantLucifer) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/exception/JdbcConnectorErrorCode.java]: return description

**Review by TyrantLucifer — CHANGES_REQUESTED** [review]

**Comment by TyrantLucifer** [comment]
@ic4y PTAL.

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSinkWriter.java [hunk 7]
```diff
@@ -85,7 +88,7 @@ public Optional<XidInfo> prepareCommit()
                 connectionProvider.getConnection().commit();
             }
         } catch (SQLException e) {
-            throw new IOException(e);
+            new JdbcConnectorException(JdbcConnectorErrorCode.EXCEPTION, e);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSinkWriter.java]: Suggest use `SQL_COMMIT_FAILED`

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcInputFormat.java [hunk 8]
```diff
@@ -95,9 +98,9 @@ public void openInputFormat() {
 
             statement = jdbcDialect.creatPreparedStatement(dbConn, queryTemplate, fetchSize);
         } catch (SQLException se) {
-            throw new IllegalArgumentException("open() failed." + se.getMessage(), se);
+            throw new JdbcConnectorException(CommonErrorCode.READER_OPERATION_FAILED, "open() failed." + se.getMessage(), se);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcInputFormat.java]: Suggest use `GET_CONNECTION_FAILED`

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcOutputFormat.java [hunk 9]
```diff
@@ -79,9 +81,8 @@ public void open()
         throws IOException {
         try {
             connectionProvider.getOrEstablishConnection();
-        }
-        catch (Exception e) {
-            throw new IOException("unable to open JDBC writer", e);
+        } catch (Exception e) {
+            throw new JdbcConnectorException(CommonErrorCode.WRITER_OPERATION_FAILED, "unable to open JDBC writer", e);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcOutputFormat.java]: Suggest use GET_CONNECTION_FAILED

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcExactlyOnceSinkWriter.java [hunk 10]
```diff
@@ -171,7 +172,7 @@ private void beginTx() throws IOException {
         try {
             xaFacade.start(currentXid);
         } catch (Exception e) {
-            ExceptionUtils.rethrowIOException(e);
+            throw new JdbcConnectorException(JdbcConnectorErrorCode.XA_EXCEPTION, e);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcExactlyOnceSinkWriter.java]: add Info "unable to start xa transaction"

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java [hunk 11]
```diff
@@ -89,18 +92,16 @@ public void open() throws SQLException {
...
         xaConnection = ds.getXAConnection();
         xaResource = xaConnection.getXAResource();
         if (jdbcConnectionOptions.getTransactionTimeoutSec().isPresent()) {
             try {
                 xaResource.setTransactionTimeout(jdbcConnectionOptions.getTransactionTimeoutSec().get());
-            }
-            catch (XAException e) {
-                throw new SQLException(e);
+            } catch (XAException e) {
+                throw new JdbcConnectorException(JdbcConnectorErrorCode.XA_EXCEPTION, e);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java]: add info "unable to set XA transaction timeout"

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java [hunk 12]
```diff
@@ -89,18 +92,16 @@ public void open() throws SQLException {
         XADataSource ds;
         try {
             ds = (XADataSource) DataSourceUtils.buildCommonDataSource(jdbcConnectionOptions);
-        }
-        catch (Exception e) {
-            throw new SQLException(e);
+        } catch (Exception e) {
+            throw new JdbcConnectorException(JdbcConnectorErrorCode.CONNECT_DATABASE_FAILED, e);
```
Comment (ic4y) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java]: add info "unable to build XADataSource"

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcInputFormat.java (hunk 8 above)
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcInputFormat.java]: Whether the CONNECT_DATABASE_FAILED can be used

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcExactlyOnceSinkWriter.java (hunk 10 above)
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcExactlyOnceSinkWriter.java]: ok

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java (hunk 11 above)
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java]: ok

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java (hunk 12 above)
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/xa/XaFacadeImplAutoLoad.java]: ok

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSinkWriter.java (hunk 7 above)
Comment (Carl-Zhou-CN) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSinkWriter.java]: Whether the TRANSACTION_OPERATION_FAILED can be used

Comment (ic4y) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/sink/JdbcSinkWriter.java]: Yes, and then add commit failure related info

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcInputFormat.java (hunk 8 above)
Comment (ic4y) [reply seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/JdbcInputFormat.java]: ok

**Comment by Carl-Zhou-CN** [comment]
@TyrantLucifer @ic4y  PTAL.

**Review by ic4y — APPROVED** [review]
LGTM

**Comment by EricJoy2048** [comment]
Please merge branch dev.

**Comment by EricJoy2048** [comment]
@TyrantLucifer  PTAL

**Review by TyrantLucifer — APPROVED** [review]

**Pull request merged**